File: practice/data_structures/linked_lists/insert_a_node_at_the_tail_of_linked_list/insert_a_node_at_the_tail_of_linked_list.cpp
/*
  Insert Node at the end of a linked list 
  head pointer input could be NULL as well for empty list
  Node is defined as 
  struct Node
  {
     int data;
     struct Node *next;
  }
*/

Node * Insert(Node * i_head, int i_data)
{
    if (!i_head)
    {
        Node * node = new Node;
        node->data = i_data;
        return node;
    }
    
    Node * node = i_head;
    while (node->next)
    {
        node = node->next;
    }
    
    Node * n = new Node;
    n->data = i_data;
    node->next = n;
    
    return i_head;
}

File: practice/data_structures/linked_lists/insert_a_node_at_the_tail_of_linked_list/insert_a_node_at_the_tail_of_linked_list.py
"""
 Insert Node at the end of a linked list 
 head pointer input could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None):
       self.data = data
       self.next = next_node
 
 return back the head of the linked list in the below method
"""

def Insert(i_head, i_data):
    if not i_head:
        return Node(i_data)
    cp = i_head
    while cp.next:
        cp = cp.next
    n = Node(i_data)
    cp.next = n
    return i_head

